### ISSUE TITLE
Git via `libgit2/examples/lg2.c` client

### ISSUE THREAD

**Issue by personalizedrefrigerator** [holzschu/a-shell#241]
## Summary
 * `libgit2`'s `examples` directory contains a simple `git` client that seems to have more working functionality than the `dulwich` client (see https://github.com/dulwich/dulwich/issues/876 and https://github.com/dulwich/dulwich/issues/877).
 * `main(int argc, char** argv)` is in [lg2.c](https://github.com/libgit2/libgit2/blob/main/examples/lg2.c).
 * The examples [seem to be in the public domain](https://github.com/libgit2/libgit2/tree/main/examples#readme), so, hopefully, there won't be any licensing issues.

**Comment by holzschu** [comment]
So, the good news is: I've managed to compile and install this CLI client to a-Shell. The bad news is: I can't make push work:
```
$ lg2 push
Error pushing [-1] - remote authentication re
quired but no callback set
```
It must be something with the configuration but I'm stuck. The pull situation is more dire:
```
$ lg2 pull
Command not found: pull
```
Clone, add and commit were fine.

**Comment by personalizedrefrigerator** [comment]
Oh no!!! I can keep working on the Dulwich-based client.

**Comment by holzschu** [comment]
I get the same error message with lg2 on Mac, so at least it's not a porting issue.

**Comment by personalizedrefrigerator** [comment]
I think this `libgit2` pull request is relevant to getting `push` to work:
https://github.com/libgit2/libgit2/pull/5904

**Comment by holzschu** [comment]
Good find!

**Comment by holzschu** [comment]
It works, but it asks for the SSH key or username/password at each launch. It would be good to make it read these from the config file, or at least try the default option (`.ssh/id_rsa`). But it works.

**Comment by personalizedrefrigerator** [comment]
> least try the default option (.ssh/id_rsa). But it works.
If we add this, it would be best to check .ssh/id_ed25519 as well -- [I think it is the recommended type of key pair to generate](https://docs.github.com/en/github/authenticating-to-github/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent).

**Comment by holzschu** [comment]
You're right. From my readings, there are no ways to specify in `.gitconfig` the key that should be used for a specific host. One possibility is to loop over all keys in `~/Documents/.ssh`, and check if one works.

**Comment by personalizedrefrigerator** [comment]
It looks like that might be doable here: 
https://github.com/libgit2/libgit2/blob/868f4bcb4d3290f4b5320f030fccdf1e7fc8ac8a/examples/common.c#L210

**Comment by holzschu** [comment]
Well, that was not completely satisfying. There is no easy way with libgit2 to check that a specific SSH key works with a given host, and no way specify which key is associated with which host (at least not without opening a session). I settled for adding an option in the `.gitconfig` file, `user.identityFile`.

**Comment by personalizedrefrigerator** [comment]
> I settled for adding an option in the .gitconfig file, user.identityFile.

https://github.com/holzschu/libgit2/pull/1 should also make this easier to use... Among other things, it tells users how to add the path to an `.ssh` key to their current repository's configuration (using `lg2 config`, which I don't think supports `--global` yet).

**Comment by personalizedrefrigerator** [comment]
Thank you for helping me with this!
Some issues I'm noticing:
 * 'lg2 status' is very slow on iOS.
     * 'dulwich status' was also very slow.
     * I don't know how to fix this...
 * I don't see 'lg2 branch' in build 172.
 * 'git log' logs **everything**
     * One solution could be to have a default number of commits to log (and a command-line option that allows the user to log more).

**Comment by holzschu** [comment]
> I don't see 'lg2 branch' in build 172.

yes, there is roughly a 12 to 24 hours delay between the time I start uploading a build to Apple servers and the time it appears on TestFlight (time to upload plus time to be approved). Build 172 corresponds to the state of the repositories on Friday night European time.

**Comment by personalizedrefrigerator** [comment]
> yes, there is roughly a 12 to 24 hours delay between the time I start uploading a build to Apple servers and the time it appears on TestFlight (time to upload plus time to be approved). Build 172 corresponds to the state of the repositories on Friday night European time.

I was basing it on the description of the build in TestFlight :). I was worried there might have been a configuration/build issue (e.g. I forgot to 'git add lg2.c' after adding the branch command).

Thank you for looking into it!

**Comment by holzschu** [comment]
Thanks for the heads up. I realized than `branch.c` is not included in the list of C files in `build_lg2.sh`. I have added it, the next build should have branches.

**Comment by personalizedrefrigerator** [comment]
I just found this (huge) PR: https://github.com/libgit2/libgit2/pull/5507

It adds a command-line `git` client to `libgit2`. From an initial look at the changes, it appears to have more features/functionality than `lg2` (but I'm not sure).

**Comment by macOneOone** [comment]
Hello @holzschu ,

Thank for all The effort you gave to this Project.

I would like to ask if you’ve any fix for this issue?
